Add bigint type

littledan opened the pull request:
This patch adds a bigint type to WebIDL with the following properties:
- bigint corresponds directly to BigInt
- The conversion from JS is based on ToBigInt, namely it throws on Number.
- bigint is not grouped as a "numeric" type, as this category is often
  used to refer to types which have a JavaScript representation of Number.
- bigint is included in the overloading resolution algorithm, and it's
  possible to overload a function for BigInt and numeric arguments.

Although it may be on the early side for such a change, the lack of
BigInt WebIDL integration is already coming up in some designs for integrating
BigInt with the Web Platform, such as https://github.com/w3c/IndexedDB/pull/231


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/littledan/webidl/pull/525.html" title="Last updated on Nov 5, 2020, 4:57 PM UTC (4f8357a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/heycam/webidl/525/d87d389...littledan:4f8357a.html" title="Last updated on Nov 5, 2020, 4:57 PM UTC (4f8357a)">Diff</a>

bzbarsky commented:
It's a little weird that this setup would allow passing `"1"` as a bigint but not allow passing `1`.

I assume bigint should not in fact be a "JSON type" because it can't be represented in JSON?  Looks to me like ToJSON throws on BigInts in the BigInt proposal.

domenic commented:
In particular this situation is interesting:

```webidl
interface X {
  void foo(bigint x);
  void bar(unsigned long long y);
}
```

What does calling `someX.foo("1")` do? What *should* it do?

bzbarsky commented:
> What does calling someX.foo("1") do?

The way the PR is written right now, it would call the `unsigned long long` overload, if you meant to name both your operations the same.  If you really did mean different names, then it would convert the string to bigint and call the one thing named "foo".

> What should it do?

Good question.  ;)

Here's a worse case:

    interface X {
      void foo(bigint x);
      void bar(bigint x);
      void bar(Node x);
    }

Here `foo("1")` would convert `"1"` to bigint, but `bar("1")` would throw, given the bits in the PR at the moment, because there is no case in https://heycam.github.io/webidl/#dfn-overload-resolution-algorithm step 12 that supports a declared type of bigint with an arg that does not have `Type(arg) == BigInt`.  So the exact changes we want to the overload resolution algorithm here require some thought.  Adding a new overload should generally not cause previously-valid calls to throw.

Also, we need corresponding changes to <https://heycam.github.io/webidl/#es-to-union>; those aren't in the PR afaict.

littledan commented:
> Also, we need corresponding changes to https://heycam.github.io/webidl/#es-to-union; those aren't in the PR afaict.

Added union support. I also added union support for Symbol (since I didn't see it added yet, and there are some funny cases where it could intersect).

bzbarsky commented:
Please don't do anything with Symbol here.  It's not in the algorithm there on purpose, because you can never have a union accepting a Symbol anyway.  The only way to pass Symbols in is via `any`.  See discussion in https://github.com/heycam/webidl/pull/377

littledan commented:
@bzbarsky Reverted `symbol` changes and added support in the overload algorithm for falling back to a `bigint` overload as the second-to-last resort (before an `any` overload).

tobie commented on index.bs:
```diff
@@ -6864,6 +6913,9 @@ ECMAScript value type.
         to an {{unrestricted double}}.
+    1.  If <a abstract-op>Type</a>(|V|) is [=BigInt=], then
+        return the result of <a href="#es-to-bigint">converting</a> |V|
+        to an {{bigint}}.
```
s/an/a/ or s/an/an IDL/

littledan commented:
> What does calling someX.foo("1") do? What should it do?

I'd say, call the numeric overload over the bigint one. Doesn't this situation already exist in cases like the following, which would also call the numeric overload over the boolean one given a string argument?

```
interface X {
  void foo(boolean x);
  void foo(unsigned long long y);
}
```

bzbarsky commented:
> I'd say, call the numeric overload over the bigint one

Why?  The bigint conversion is the less lossy one in this case, I would think.

> which would also call the numeric overload over the boolean one given a string argument?

It would, yes.  Arguably conversion to numeric is less lossy than conversion to boolean.

More generally, it seems we're going out of our way to provide seamless conversions from strings (but not other primitives!) to bigints.  This is in contrast to numeric and boolean types, which can be converted from all primitives.  So in some sense, there is some sort of privileged relationship between bigint and strings that is not shared by other primitive values in JS or other primitive-like idl types...

bzbarsky commented:
Oh, and unlike a lot of overload scenarios, the "overload numeric and bigint" is not completely insane, because if you want to have an API that takes "numbers of arbitrary precision" your options are:

1. Have it just take `bigint`.  Callers can pass `BigInt(0)` or `"0"` but not `0`.
2. Have it take `(bigint or some-numeric-type)`.  Callers can pass `0` sanely.

Option 1 is slightly harder to use as an API, and the reasoning behind `"0"` being OK but `0` not being OK is subtle and likely to make anyone trying to just pass a small numeric constant be annoyed.

Option 2 involves an overload of bigint and numeric, and dataloss if you pass a long-enough string and it gets converted to the numeric.

littledan commented:
@bzbarsky I see; maybe we should then prohibit that particular overload and force an explicit string overload as well? There could still be some more obscure data loss cases, though.

bzbarsky commented:
There really isn't a mechanism so far for prohibiting an overload but allowing it of _more_ types are added.  We could make bigint and numeric not distinguishable, forcing option 1.  Or we could add some sort of mechanism along these lines.  Or we could make es-to-idl conversion for bigint more like https://tc39.github.io/proposal-bigint/#sec-bigint-constructor-number-value than like https://tc39.github.io/proposal-bigint/#sec-to-bigint I suppose....

littledan commented:
> Oh, and unlike a lot of overload scenarios, the "overload numeric and bigint" is not completely insane, because if you want to have an API that takes "numbers of arbitrary size" 

I'm not sure if that's in line with the goals of the BigInt proposal, to have new APIs which take both out of the box by design. I think most APIs fall into the category of "logically making sense for arbitrary-size integers" or "logically making sense for floating point values"; there's definitely stuff in the intersection ("logically making sense for only small integers"), and for that, Number works well. I was imagining overloading would mostly make sense when trying to upgrade an API from a Number version to a BigInt version, when we initially thought that it made sense only on small integers, but it turned out it was also important on large integers.

> Option 1 is slightly harder to use as an API, and the reasoning behind "0" being OK but 0 not being OK is subtle and likely to make anyone trying to just pass a small numeric constant be annoyed.

The hypothesis of the BigInt proposal design is that programmers will keep track of whether something is a BigInt operation or a Number operation, and the types of scalar values, and not do things generically. All the JS programmers I've heard from seem basically on board with this design. Part of this here is, even for small numeric constants, you put an `n` at the end of the literal. I don't want to encourage a proliferation of Web APIs which 

> Or we could make es-to-idl conversion for bigint more like https://tc39.github.io/proposal-bigint/#sec-bigint-constructor-number-value than like https://tc39.github.io/proposal-bigint/#sec-to-bigint I suppose....

If we did that, and privileged the bigint path, we'd get surprising exceptions for things like `f("3.2")`, wouldn't we?

I don't see a great answer here which makes everything beautiful, intuitive and not throw errors. Maybe we should start by making `bigint` and `number` indistinguishable, and make sure that unions throw in the ambiguous case. I uploaded a patch which does that.

bzbarsky commented:
> Part of this here is, even for small numeric constants, you put an n at the end of the literal.

Ah, if there are simple bigint literals that helps a lot.

> Maybe we should start by making bigint and number indistinguishable

Yes, please.

littledan commented:
> Yes, please.

How does [this patch](https://github.com/heycam/webidl/pull/525/commits/fa92cf3a23598a654df0f563921cebe0578d94fc) look?

inexorabletash commented:
Preface: this is all excellent discussion and @littledan's proactive approach to getting bigint integrated into the platform is awesome.

> Although it may be on the early side for such a change...

On that note: are there concrete examples of API that plan to consume/produce bigints, either via WebIDL or ES? (Indexed DB is unfortunately not a good example, since it requires custom integration)

I'm slightly nervous that since it's too early to have good examples of bigint usage we may be missing something. @littledan describes "the hypothesis of the BigInt proposal design" which makes sense to me wanting to keep mentally separate from the rest of the platform.... but what will eventual idiomatic use of bigints look like?

littledan commented:
WebIDL has a number of cases where it switches among all of the JS primitive types. If BigInt is a type in JavaScript, maybe we should have some description or other of what the behavior is when a BigInt ends up floating into certain operations, such as [converting an ES value to any](https://heycam.github.io/webidl/#es-any), even if we want to wait until we have more information to decide how to define what it means to overload with a `bigint` type.

One place BigInt will come up is the WebAssembly-JS API. If the [proposed BigInt/i64 integration](https://github.com/WebAssembly/spec/pull/707) is combined with the [proposed WebAssembly.Global](https://github.com/WebAssembly/mutable-global) interface, then I think we'd want to say that the type of the WebAssembly.Global.prototype.valueOf method is `bigint or unrestricted double`, for example. But maybe we could define something special in prose in that particular document, without updating WebIDL so generally.

Besides that, I don't have a particular other web platform feature that where I want to propose BigInt usage. All the examples I can think of (constant time 64-bit arithmetic operations, a date format representing time in nanoseconds since forever ago, etc) might make more sense in the JavaScript standard itself than as WebIDL-based APIs. Like the IndexedDB integration, I'm fine waiting until we have more motivation for this change.

As much as it's appropriate to wait for more applications to get input for the design of the BigInt WebIDL binding, I wanted to document a suggestion for best practices here, at least from the perspective of the way BigInt was designed in JavaScript, in particular, avoiding implicit conversions that lose information, and continuing to use Number exclusively everywhere it is currently used in WebIDL.

bzbarsky commented:
> The only way to pass Symbols in is via any

This may not be true.  I am checking with Domenic.

`any` definitely needs to handle all JS types, so we need something there even if we don't add a syntactic way to declare bigints.

littledan commented:
Another way to handle both the union and overloading semantics could be to do something similar to what the JS spec does: Call ToNumeric in the ambiguous cases, and make the judgment based on whether a Number or BigInt falls out. This is what we do for most overloaded operators such as `-`. ToNumeric converts strings to Numbers--this is required, for compatibility. Presumably a similar compatibility risk (but perhaps not as serious, depending how much the method is used) would exist if there's an existing method which takes a numeric type and wants to add a bigint overload, or a bigint member to a union.

littledan commented:
Any thoughts on whether we should support overloading between BigInt and Number or not? Note that one specification proposal where this arises is the integration of BigInt and Intl.NumberFormat (not implemented anywhere AFAIK): https://github.com/tc39/ecma402/pull/236 . That PR follows the ToNumeric algorithm as in https://github.com/heycam/webidl/pull/525#issuecomment-372964767 .

annevk commented:
I think it actually makes sense to not allow overloading/union BigInt with any other IDL numeric type as that allows for (maybe even encourages) the kind of APIs TC39 doesn't want. Where the API silently does BigInt(arg) if you pass it a non-BigInt after crossing the IDL boundary.

I'm also not sure why we'd support coercing string to BigInt, I think it makes sense to throw there as well.

annevk commented:
What does JavaScript do for its methods that take a BigInt?

littledan commented:
Well, the first built-in method proposed (not yet shipped anywhere AFAIK) is  tc39/ecma402#236 which works like operators like `*` and calls ToNumeric on the argument (which will convert a string to Number, or `{valueOf(){return 5n}}` to `5n`), then checks whether it's Number or BigInt. That's what I was suggesting in https://github.com/heycam/webidl/pull/525#issuecomment-444073662 , but if these semantics seem terrible, we could go with something more restrictive since AFAIK no overloaded methods have shipped, just overloaded operators (which have more legacy constraints).

annevk commented:
I think that's fine as long as we throw a TypeError if it turns out to be a Number and don't allow overloading/unioning a BigInt with a Number.

littledan commented:
OK, so your suggestion would be to say, the infix arithmetic operations are the only things overloaded, and otherwise we use different names?

annevk commented:
I thought the infix arithmetic operations were not overloaded? Beyond valueOf() that is?

I'd like to disallow
```webidl
a(double x);
a(BigInt x);
b((double or BigInt) x);
```
as per my understanding we don't want implicit `BigInt(x)`. However, if we have
```webidl
c(BigInt x);
```
and that does `ToNumber()` first for `valueOf()` similar to the infix arithmetic operations I think that'd be fine and nicely consistent.

littledan commented:
Infix arithmetic operations work on BigInt and Number, but both operands need to be either BigInt or Number, not mixed.

Overloading like 
```
a(double x);
a(bigint x);
```
would not cause an implicit conversion from Number to BigInt; it would just call `ToNumeric` on the argument, which would call `valueOf` on objects, and convert strings to Number, and leave Numbers and BigInts in place. That's just like what arithmetic operations like unary `-` do--still no implicit conversion from Number to BigInt, but also keeping in place all the existing conversions.

If you define `c(bigint x);`, this will call ToNumeric, and which calls valueOf but doesn't convert a Number to a BigInt; it will just throw on Number. (That's the case whether or not we allow the overloading.)

The patch currently out for review permits unions like `(double or bigint)` (I don't see any particular way to prohibit it), but if it's checking for membership of a numeric value in a union like that, it will throw an error.

annevk commented:
Right, so I'm saying we should try to prevent such overloading and unions (perhaps with a prose restriction) as otherwise they'll be used to do the kind of thing we don't want people to be doing.

littledan commented:
I see, sounds like we should change the Intl patch if we want to go this way.

littledan commented:
We discussed the overloading issue in TC39 and concluded that `Intl.NumberFormat.prototype.format` *should* overload between Number and BigInt (based on calling ToNumeric and then checking whether a Number or BigInt came out). So, we may want to permit that here, or we could still prohibit it. Any preferences? Based on a ping from @cynthia about TAG interest, I'd like to conclude on these issues land this PR soon.

domenic commented:
It seems like it would be easier to judge if there were web APIs wanting to use bigint. Are there any currently?

annevk commented:
Well, `Intl.NumberFormat.prototype.format`, per "jsidl".

We also have a case that shows that allowing a union/overload will likely lead to APIs with implicit coercion of Number to BigInt, which is something TC39 does not want, as far as I know: https://github.com/WebAssembly/JS-BigInt-integration/issues/12.

I don't think there's a way to enforce the difference through IDL. I guess you could try something along the lines of that when there's a union, the corresponding algorithm cannot contain a Number to BigInt operation, but try enforcing that realistically …

Ms2ger commented:
Looking into this some more, adding support for unions that include bigint and a numeric type (with the "call `ToNumeric` and see what falls out" semantics) is fairly straightforward. Adding support for overloading is significantly more hairy, especially when involving unions (such as in the example below). The only real use case I've seen so far is [`NumberFormat`](https://github.com/tc39/ecma402/pull/236), and unions are sufficient there. My inclination would be not to allow overloading at this point; I'm fine going either way for unions – we could discuss again if jsidl ever happens.

```idl
interface X {
  void foo((bigint or Node) x);
  void foo(long x);
};
```

Thoughts?

annevk commented:
I'm okay with allowing unions, with a warning to ask for advice (prolly filing an issue against this repository) and a high-level outline of the principles when it is and when it is not appropriate.

Ms2ger commented:
Updated some more; we can look into renaming "numeric types" separately.

annevk commented on index.bs:
```diff
@@ -6529,6 +6579,19 @@ Each pair of [=flattened member types=]
+precision errors.
+Please
+<a href="https://github.com/heycam/webidl/issues/new?title=Intent%20to%20use%20bigint/numeric union">file an issue</a>
+to discuss before using this feature.
```
```suggestion
before using this feature.
```

bathos commented on index.bs:
```diff
@@ -5768,6 +5807,7 @@ type.
         "boolean"
         "byte"
         "octet"
+        "bigint"
```
The addition of this new unnamed terminal means the `Other` production needs to be updated to include that terminal here https://github.com/heycam/webidl/blob/c387fed823d2d4ccf8940bbc065f358ccf563073/index.bs#L6998-L6999

(Unnamed/quoted terminals ‘delete the token’ from the set of values represented by a named terminal, so _ident_ is currently inclusive of the string "bigint," but will cease to be, hence the need to explicitly add it to Other, which is inclusive of every terminal except for a few specific punctuation marks.)

bathos commented on index.bs:
```diff
@@ -14555,6 +14682,11 @@ expression syntax [[!PERLRE]]) as follows:
+    <tr>
+        <td id="prod-bigint"><emu-t class="regex">bigint</emu-t></td>
+        <td><code>=</code></td>
+        <td><code class="regex"><span class="mute">/</span>-?([1-9][0-9]*|0[Xx][0-9A-Fa-f]+|0[0-7]*)n<span class="mute">/</span></code></td>
```
The combined lexical goal can currently be realized case insensitively (assuming one handles unnamed terminals in a lookup step). That will cease to be true after this change because this "n" is case sensitive. This isn’t a problem, but I’m mentioning it because it could plausibly impact existing Web IDL lexers if they’ve made certain assumptions.

Ms2ger replied:
I don't think we actually need a bigint token; integer works fine. I did that instead.

bathos commented on index.bs:
```diff
@@ -5892,6 +5932,16 @@ The [=type name=] of the
 {{octet}} type is "<code>Octet</code>".
 
 
+<h4 id="idl-bigint" interface>bigint</h4>
```
This is placed between `octet` and `short` here in the Types list. Later in the ECMAScript type mapping section it’s placed between `unrestricted double` and `DOMString`. These lists currently use the same ordering and I think that should probably be preserved since it makes things a bit easier to follow. The latter positioning makes sense. Slotting it between two of the sized integer types seems strange:

- ± 8
- \+ 8
- ± ∞
- ± 16
- \+ 16
- ± 32
- \+ 32
- ± 64
- \+ 64

Ms2ger commented:
I think I addressed the remaining comments. @bzbarsky, would you mind reviewing?

bzbarsky commented:
I will try to get to this, but there's a lot to page in and I am pretty swamped with other things at the moment.  I can't promise any timeline other than "not in the next two days"...

annevk commented:
So IDB doesn't depend on this as far as I can tell and it's not clear that's moving forward anyway. Are there other places that might use this through IDL? I'd rather not add it until we have a concrete consumer I think.

ExE-Boss commented:
Well, **WebAssembly** intends to use `bigint` for `i64` handling in **JavaScript** code: <https://github.com/WebAssembly/spec/pull/707>.

Currently, the `value` attribute type of `WebAssembly.Global` is `any`, because no other **WebIDL** type currently allows **bigint** values.

Ms2ger commented:
I guess the way WebAssembly uses `any` is problematic wrt the formal definition of `any` as a discriminated union, and fixing this would require a bigint type in IDL. Note future changes in the [reference-types](https://webassembly.github.io/reference-types/js-api/index.html#tojsvalue) proposal will extend `WebAssembly.Global` with a bunch of non-integer types.

annevk commented:
Ah, so that would be a problem for `postMessage()` and such too then? That does seem like something we ought to fix. (The weird thing with WebAssembly is that browsers don't seem to use IDL to implement it.)

littledan commented:
Right, if we don't want to permit bigint as an actual IDL type, we should make certain changes to WebIDL to let it "pass through". I've been hoping that this patch would land, so I didn't formulate another one for that subset. (I think WebIDL is just sort of incoherent right now, with BigInt existing as a type in JS.)

I've heard from a few people in the TAG that they are interested in this patch landing. Comments here about why would be appreciated. @cynthia @kenchris

If we *conclude* that this patch won't land, then let's formulate that more minimal patch and land that instead.

rtoy commented:
It would make the [API for a WebAudio noise generator](https://github.com/WebAudio/web-audio-api-v2/issues/8) a bit nicer if BigInt were supported.  We currently using an array of two 32-bit unsigned ints to specify the 64-bit seed we need.

domenic commented:
@rtoy the only thing this PR was waiting on was an actual consumer, so if you're providing one, then I think we should go ahead and merge this. @annevk?

annevk commented:
That seems reasonable. And it doesn't seem to want to use overloading so we don't have to address that for now.

rtoy commented:
Just to follow up, see https://github.com/WebAudio/web-audio-api-v2/issues/8#issuecomment-721439274 and https://github.com/WebAudio/web-audio-api-v2/issues/8#issuecomment-721803254.  Using BigInt makes the API a bit nicer, but an array of 2 32-bit unsigned integers works fine too.

domenic reviewed (approved):
I just did a re-review and this looks good to me. @TimothyGu, would you be able to give it a double-check and/or merge?

TimothyGu commented:
CI seems to be failing due to infrastructural reasons.
